File: leetcode/547. Number of Provinces.cpp

#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    int n;
    vector<bool> visted;

    void dfs(int node, vector<vector<int>> &table){
        visted[node] = true;
        for (int i=0; i<n; i++){
            if (!visted[i] and table[node][i] == 1) dfs(i, table);
        }
    }

    int findCircleNum(vector<vector<int>>& isConnected) {
        n = isConnected.size();
        visted.resize(n, false);
        int ans = 0;
        for (int i=0; i<n; i++){
            if (!visted[i]){
                ans += 1;
                dfs(i, isConnected);
            }
        }
        return ans;
    }
};

File: leetcode/547. Number of Provinces.py
from typing import List

class Solution:
    def dfs(self, node: int) -> None:
        self.visted[node] = True
        for i, status in enumerate(self.table[node]):
            if (status == 1 and self.visted[i] is False):
                self.dfs(i)

    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        self.n = len(isConnected)
        self.visted = [False for _ in range(len(isConnected))]
        self.table = isConnected

        ans = 0
        for i in range(self.n):
            if self.visted[i] is True:
                continue
            ans += 1
            self.dfs(i)
        
        return ans